receiver/k8sobjectsreceiver/receiver.go

```diff
@@ -148,36 +151,67 @@ func (kr *k8sobjectsreceiver) startPull(ctx context.Context, config *K8sObjectsC
 }
 
 func (kr *k8sobjectsreceiver) startWatch(ctx context.Context, config *K8sObjectsConfig, resource dynamic.ResourceInterface) {
-
 	stopperChan := make(chan struct{})
 	kr.mu.Lock()
 	kr.stopperChanList = append(kr.stopperChanList, stopperChan)
 	kr.mu.Unlock()
 
-	resourceVersion, err := getResourceVersion(ctx, config, resource)
-	if err != nil {
-		kr.setting.Logger.Error("could not retrieve an initial resourceVersion", zap.String("resource", config.gvr.String()), zap.Error(err))
-		return
-	}
 	watchFunc := func(options metav1.ListOptions) (apiWatch.Interface, error) {
 		options.FieldSelector = config.FieldSelector
 		options.LabelSelector = config.LabelSelector
 		return resource.Watch(ctx, options)
 	}
 
+	cancelCtx, cancel := context.WithCancel(ctx)
+	cfgCopy := *config
+	wait.UntilWithContext(cancelCtx, func(newCtx context.Context) {
+		resourceVersion, err := getResourceVersion(newCtx, &cfgCopy, resource)
+		if err != nil {
+			kr.setting.Logger.Error("could not retrieve a resourceVersion", zap.String("resource", cfgCopy.gvr.String()), zap.Error(err))
+			cancel()
+			return
+		}
+
+		done := kr.doWatch(newCtx, &cfgCopy, resourceVersion, watchFunc, stopperChan)
+		if done {
+			cancel()
+			return
+		}
+
+		// need to restart with a fresh resource version
+		cfgCopy.ResourceVersion = ""
+	}, 0)
+}
+
+// doWatch returns true when watching is done, false when watching should be restarted.
+func (kr *k8sobjectsreceiver) doWatch(ctx context.Context, config *K8sObjectsConfig, resourceVersion string, watchFunc func(options metav1.ListOptions) (apiWatch.Interface, error), stopperChan chan struct{}) bool {
 	watcher, err := watch.NewRetryWatcher(resourceVersion, &cache.ListWatch{WatchFunc: watchFunc})
 	if err != nil {
 		kr.setting.Logger.Error("error in watching object", zap.String("resource", config.gvr.String()), zap.Error(err))
-		return
+		return true
 	}
 
+	defer watcher.Stop()
 	res := watcher.ResultChan()
 	for {
 		select {
 		case data, ok := <-res:
+			if data.Type == apiWatch.Error {
+				errObject := apierrors.FromObject(data.Object)
+				var statusErr *apierrors.StatusError
+				// FromObject guarantees an unwrapped StatusError or UnexpectedObjectError so this type checking is best
+				// nolint:errorlint
+				statusErr, ok = errObject.(*apierrors.StatusError)
+				if ok && statusErr.ErrStatus.Code == http.StatusGone {
```

Comment:
If it's guaranteed, can we avoid the check completely? If it'll fail at any point in future, it'll be very hard to identify. So, we need to log an explicit error or panic.
```suggestion
				if errObject.(*apierrors.StatusError).ErrStatus.Code == http.StatusGone {
```

Reply:
Updated.